### PR TITLE
Add `asSubtype`

### PR DESCRIPTION
**Pull request by alita-moore**
I frequently run into edgecases that typescript has not addressed in their auto labeling of types. So I'm instead forced to redefine types or unnecessarily address supposedly nullable variables. This is a function to make overcoming these limitations easier.

### CODE REVIEW

File: source/as-subtype.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
+/**
+Cast the type of a variable to a subtype.
+
+This is useful in cases where the typescript engine is unable to catch certain implicit assertions.
```
Comment (sindresorhus) [inline source/as-subtype.ts:4]: What cases? Can you describe succinctly and example situation where it might be needed?

Comment (sindresorhus) [inline source/as-subtype.ts:4]:
```suggestion
This is useful in cases where TypeScript is unable to catch certain implicit assertions.
```

File: source/as-subtype.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+const baz = "baz" as string | null;
+asSubtype<string>(baz)
+//=> string
+asSubtype<null | string>(baz)
+//=> string
+```
+ */
+export function asSubtype<T>(value: any): asserts value is T {
+	if (!(value || !value)) {
+		console.log('this is to assuage the linter');
```
Comment (sindresorhus) [inline source/as-subtype.ts:57]: I think it's better to use an ESLint ignore comment.

File: test/as-subtype.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
+import test from 'ava';
+import {expectTypeOf} from 'expect-type';
+import {asSubtype} from '../source/index.js';
+
+test('isDefined()', t => {
```
Comment (sindresorhus) [inline test/as-subtype.ts:5]: Typo

File: test/as-subtype.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+	// To assuage the linter
+	t.true(true);
```
Comment (sindresorhus) [inline test/as-subtype.ts:7]:
```suggestion
	t.pass();
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/as-subtype.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+Cast the type of a variable to a subtype.
+
+This is useful in cases where the typescript engine is unable to catch certain implicit assertions.
+
+If a impossible type conversion is performed the result will be never.
+
+You cannot expand a type; you may only reduce or translate to a similar type.
+
+@example
+```
```
Comment (sindresorhus) [inline source/as-subtype.ts:11]:
```suggestion
```
import {asSubtype} from 'ts-extras';
```

File: source/as-subtype.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+// This is particularly useful in async actions and
```
Comment (sindresorhus) [inline source/as-subtype.ts:12]:
```suggestion
import {asSubtype} from 'ts-extras';

// This is particularly useful in async actions and
```

File: source/as-subtype.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+// You can only increase precision
+const baz = 'baz' as string | null;
+asSubtype<string>(baz)
+//=> string
+asSubtype<null | string>(baz)
+//=> string
+```
+*/
+export function asSubtype<T>(value: any): asserts value is T {
```
Comment (sindresorhus) [inline source/as-subtype.ts:55]:
```suggestion
export function asSubtype<T>(value: unknown): asserts value is T {
```

**Comment by sindresorhus** [comment]
@alita-moore Friendly bump :)

**Comment by alita-moore** [comment]
Thanks! I'll address all comments on my PRs this weekend!

**Comment by sindresorhus** [comment]
Great :)

Note that you now also need to add the types to the readme and add appropriate categories to the TS doc comments.

**Comment by sindresorhus** [comment]
Bump :)

**Pull request closed without merging**